zimcheck complains about percent-encoded URLs

URLs like `path/%26` are valid, aren’t they? When creating ZIM files for the DWDS project, we have files named like `&` (a lonesome ampersand) or `'` (apostrophe), and the right way to link to them seems to be `<a href="path/%26">…</a>`, `<a href="path/%27">…</a>`.

Minimal test case:

```
% cat test.html
<a href="%26">&amp;</a>
<a href="%27">&apos;</a>

% echo foo > "&"

% echo foo > "'"
```

After creating a ZIM file `test.zim` with `zimwriterfs`, the check via `zimcheck test.zim` fails with:

```
[ERROR] Invalid internal links found:
  The following links:
- %26
- %27
() were not found in article test.html
```

--- a/src/tools.cpp
+++ b/src/tools.cpp
@@ -450,8 +450,10 @@ std::string normalize_link(const std::string& input, const std::string& baseUrl)
 
         if ( *p == '%')
         {
-            if( (p+3) >= input.cend()){
+            if( (p+2) >= input.cend()){
                 // if the %XX token would go off the end of the string, just break
+                // string::end() returns an iterator pointing to the null terminator
+                // of the underlying c-string (guarenteed as of C++11)
                 break;
             }
             // hhx only officially supports hex unsigned char

--- a/test/tools-test.cpp
+++ b/test/tools-test.cpp
@@ -260,6 +260,9 @@ TEST(tools, normalize_link)
     ASSERT_EQ(normalize_link("%", "/"), "/");
     ASSERT_EQ(normalize_link("%1", ""), "");
 
+    EXPECT_EQ(normalize_link("%26", ""), "&");
+    EXPECT_EQ(normalize_link("%27", "/"), "/\'");
+
     // ../test/tools-test.cpp:260: Failure
     // Expected equality of these values:
     //   normalize_link("%", "/")

--- a/test/tools-test.cpp
+++ b/test/tools-test.cpp
@@ -260,8 +260,8 @@ TEST(tools, normalize_link)
     ASSERT_EQ(normalize_link("%", "/"), "/");
     ASSERT_EQ(normalize_link("%1", ""), "");
 
-    EXPECT_EQ(normalize_link("%26", ""), "&");
-    EXPECT_EQ(normalize_link("%27", "/"), "/\'");
+    ASSERT_EQ(normalize_link("%26", ""), "&");
+    ASSERT_EQ(normalize_link("%27", "/"), "/\'");
 
     // ../test/tools-test.cpp:260: Failure
     // Expected equality of these values:

